Logout button to profile page + profile refactor + loading button

gruvw opened the pull request:
Fixes #134 

This PR aims to move the logout button from its temporary position on the home page to the profile page.
In order to do, so I had to heavily refactor the Profile Page and apply a slight change to our error handling process.

**Acceptance criteria**:

- [X] No logout button on the home page
- [X] See a logout button on profile page
- [x] Clicking on the logout button on the profile page redirects to the login page without errors (because there was one at first).
- [X] The deletion button should be able to accept an arbitrary future as parameter and display a spinner while the deletion occurs.

**Result**: profile page app bar with logout button

![image](https://github.com/ProximaEPFL/proxima/assets/63407038/9cc9dfbf-8e4d-4aca-ad85-be2f538e2593)

see https://github.com/ProximaEPFL/proxima/pull/148#issuecomment-2076904196 for the loading deletion button.

**Note**: this PR also bumps the project version according to MS2 sprint 2.

**Note**: waiting for #144 PR to be merged.

yoannLafore commented on lib/views/pages/profile/info_cards/profile_info_pop_up.dart:
```diff
@@ -0,0 +1,74 @@
+
+  final String? title;
+  final String content;
+  final VoidCallback onDelete;
```
The `onDelete` function is probably going to invoke some repository calls. 
Thus, this would mean that it returns a `Future<void>`.

So, I wonder if we shouldn't replace the type of `onDelete` as below:
```suggestion
  final Future<void> Function() onDelete;
```

And then we `await` it during the `onPressed` before navigating back.
What do you think ?

(If you choose to go with that, I believe you can apply this change in `profile_info_card` as well)

gruvw replied:
Yes I am on it, this requires a few changes. I'll keep you posted.

gruvw replied:
see PR comments

yoannLafore reviewed (commented):
Great work :tada: 
Thank you for refactoring the profile page, the code is now much better organized and I will definitely take inspiration from your work in the future.
In particular, good idea to make some parameters (like `title` in the `ProfileInfoCard`) optional, this helps to reduce code duplication :+1: 

I have just made small suggestions below (mostly style).
Other than that, the PR looks good to me.

Thank you for your work !

gruvw commented on lib/views/components/logout_button.dart:
```diff
@@ -0,0 +1,18 @@
+  Widget build(BuildContext context, WidgetRef ref) {
+    return IconButton(
+      key: logoutButtonKey,
+      onPressed: () => ref.read(loginServiceProvider).signOut(),
```
Haha yes nice catch ;)

Fixed in a911468

gruvw commented on lib/views/pages/profile/components/user_account.dart:
```diff
@@ -0,0 +1,57 @@
+      overflow: TextOverflow.fade,
+    );
+
+    //TODO: get the user points from the viewmodel
```
Yes, fixed by 946ef53

gruvw commented on lib/views/pages/profile/components/user_account.dart:
```diff
@@ -0,0 +1,57 @@
+    const profilePicture = CircleAvatar(
+      radius: 20,
+      //TODO: get the image from the viewmodel
+      //backgroundImage: AssetImage("assets/images/user.png" ),
```
I think we might want to keep this as it could make sense when we will implement profile pictures right ?
What do you think ?

yoannLafore replied:
I didn't saw it like that but effectively we can keep it then :+1:

camillelnne commented on lib/utils/ui/circular_value.dart:
```diff
@@ -12,8 +12,11 @@ class CircularValue<T> extends StatelessWidget {
-      const SizedBox.shrink();
+  /// Tag to be placed at the start of an error message to display the
+  /// message directly instead of the widget.
+  /// Does not trigger a pop up dialog. Usefull for debug error messages
```
Minor typo
```suggestion
  /// Does not trigger a pop up dialog. Useful for debug error messages
```

gruvw replied:
sure, done in 505f0e6

camillelnne commented on lib/views/pages/home/top_bar/app_top_bar.dart:
```diff
@@ -21,7 +19,7 @@ class AppTopBar extends HookConsumerWidget {
+    final userAvatar = CircleAvatar(
       child: Stack(
         children: [
           const Center(child: Text("PR")),
```
```suggestion
          const Center(child: Text("PR")), //TODO: get real initials
```

Implementing it is not in the scope of your PR, but I think it would be good to have the correct initials.
Could add a TODO so we don't forget to fix it? (and I can create an issue in the backlog).
Or quick fix it maybe, if it's easier for you.
@yoannLafore, as scrum master, what do you think?

gruvw replied:
I'll fix it

gruvw replied:
Yeah, never mind this has nothing to do with this PR in reality, I misunderstood the problem you were pointing at first.

I will create an issue in the backlog about have a clear component for the user profile avatar that we can reuse throughout the app.

gruvw replied:
see #157

camillelnne commented on lib/views/pages/profile/components/profile_badge.dart:
```diff
@@ -0,0 +1,37 @@
+class ProfileBadge extends StatelessWidget {
+  static const badgeKey = Key("profileBadge");
+
+  static const _badgeWidth = 54.0;
```
By curiosity, does this adapt well to all different phone sizes? As it is a hard coded value.

gruvw replied:
Yeah it is basically like an icon size. I don't think it's a problem to have fixed values for those, it's like padding or things like that.

camillelnne commented on lib/views/pages/profile/info_cards/profile_info_card.dart:
```diff
@@ -0,0 +1,88 @@
+  static const infoCardKey = Key("profileInfoCard");
+
+  static const _borderRadius = BorderRadius.all(Radius.circular(10));
+  static const _cardHeight = 80.0;
```
Same question for the fixed size here.

gruvw replied:
same answer

camillelnne reviewed (commented):
Awesome work!
Just made a few comments, but I think the PR is ready to be merged after @yoannLafore discussions are resolved.

gruvw commented:
Hey, so after quite some trouble, I finally managed to implement a decent async deletion button (as suggested by https://github.com/ProximaEPFL/proxima/pull/148#discussion_r1578233154).

**Note**: retrospectively, I think that this would have been better in its own PR. I updated the acceptance criteria and changed the title to accommodate this change. Not ideal, but we will make it work.

So I implemented a new component `DeleteButton` which is an async capable delete button.
It means that you can pass an async function to it, and it will execute the function once you click the button while displaying a spinner till the function is done executing.

**Result**

On Pop Up:
![Peek 2024-04-25 12-23](https://github.com/ProximaEPFL/proxima/assets/63407038/3b4d4c0e-1238-4d27-9b14-f9024e41ffcf)

On Post in profile page:
![Peek 2024-04-25 12-30](https://github.com/ProximaEPFL/proxima/assets/63407038/1ada5fa1-2e5b-446f-a7f4-1130b91ba06e)

Here I display a checkmark (when the future completes) for displaying purposes, but of course we can `.then(...)` a redirection or trigger a refresh right away in the async function that we pass.

gruvw commented:
After some thoughts, I made the delete button generic because there is no real reason to keep it only for deletion purposes.
It is now called `LoadingIconButton` and there is a child class called `DeleteButton`, see implementation:

https://github.com/ProximaEPFL/proxima/blob/4cb1c2de10d6fb38e65325c85836f27fed25277a/lib/views/components/loading_icon_button.dart#L68-L71

gruvw commented:
Ok I resolved all your comments, let me know if you need anything else or if you have any feedback on my additions.

yoannLafore commented on lib/views/components/loading_icon_button.dart:
```diff
@@ -0,0 +1,71 @@
+import "package:flutter_hooks/flutter_hooks.dart";
+import "package:hooks_riverpod/hooks_riverpod.dart";
+
+typedef FutureVoidCallback = Future<void> Function();
```
Good idea to define a `typedef` !
I was just wondering if this file was a good place to define it.
May be we should have it in a utility file instead ?

yoannLafore replied:
Yes I see.
I believe we can still add a util file even if it is one line because we might forget to refactor the current implementation in the future I feel like.

gruvw replied:
ok done in 23133e5

yoannLafore reviewed (approved):
Amazing work :tada: 
Thank you for making the delete button more reactive, I believe this greatly contribute to the user experience. It looks absolutely beautiful !

I have just let a little comment below but other than that the PR looks good to me :+1: 
Thank you for your great work !

camillelnne reviewed (approved):
Thanks for the changes and good job for the delete button!
